File: Data Structures & Algorithms/subsets-ii/submission-0.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        
        # isnt the trick the same as the Combination Sum II

        res = []
        nums.sort()
        
        def dfs(index, subset):

            #end case
            if index >= len(nums):
                res.append(subset.copy())
                return

            
            #take this
            subset.append(nums[index])
            dfs(index + 1, subset)

            #dont take this
            subset.pop()

            while (index < len(nums) - 1) and nums[index] == nums[index + 1]:
                index += 1

            dfs(index + 1, subset)

        dfs(0, [])
        
        return res

File: Data Structures & Algorithms/maximum-product-difference-between-two-pairs/submission-0.cpp
class Solution {
public:
    int maxProductDifference(vector<int>& nums) {
        sort(nums.begin(), nums.end()); int n = nums.size();

        int res1 = nums[n-1] * nums[n-2]; int res2 = nums[0] * nums[1];
        return res1 - res2;
    }
};